design-documents/storefront/catalog/product-options-and-variants.md

```diff
@@ -0,0 +1,363 @@
+
+The link on a product that exists in the same domain could be done through the shared ID, which as for me makes sense since I can see a possible scenario when variation that initially had only a price will be "promoted" to a product. For example due to integration reasons, to pass additional attributes data along with variant to Google Merchant Center.
+
+### Enhanced option values
+
+To reach a better user experience,
+the option could be extended with an image that represents it 
+or info URL that provides additional information about option value.
+
+[Magento: Swatches](https://docs.magento.com/user-guide/catalog/swatches.html)
+
+[Magento: Downloadable products, options with samples](https://docs.magento.com/user-guide/catalog/product-create-downloadable.html)
+
+Both types of resources could be specified as attributes of `ProductOptionValue`.
+
+```proto
+syntax = "proto3";
+
+message Product {
+    repeated ProductOption options = 100;
+}
+
+message ProductOptionValue {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isDefault = 4;
+    string imageUrl = 5;
+    string infoUrl = 6;
+}
+
+message ProductOption {
+    string id = 1;
+    string label = 2;
+    string sortOrder = 3;
+    string isRequired = 4;
+    string renderType = 6;
+    repeated ProductOptionValue values = 5;
+}
+
+message ProductVariant {
+    repeated string optionValueId = 1;
+    string id = 2;
+    int32 weight = 3;
+    string productIdentifierInPricing = 500; #*
+    string productIdentifierInInventory = 600; #*
+}
+#* to avoid unnecessary network calls the variant has to know does it have a link on another domain,  type, and proper name of a field representing this link TBD.
+```
+
+![](https://app.lucidchart.com/publicSegments/view/eed59f85-7d04-46ac-9d7e-eaa77073017e/image.png)
+
+### Explaining data and operations through the pseudo SQL
+
+Lets review pseudo SQL schema, which implements the picture described above.
+This is not the instruction for implementation, tables intentionally do not have all the fields of data objects.
+The example proposed to show the relations and operations that we have in the domain.
+
+```sql
+create table products (
+    object_id char(36) not null,
+    data json not null,
+    primary key (object_id)
+);
+```
+Table `products` stores registry of products.
+
+```sql
+create table product_variant_matrix (
+    value_id char(36) not null,
+    object_id char(36) not null,
+    weight tinyint not null,
+    primary key (value_id, object_id)
+);
+
+alter table product_variant_matrix
+    add foreign key fk_product_variant_matrix_value_id (value_id) references product_option_values(value_id);
+```
+
+`product_variant_matrix` - Stores the correlation between option values and variants, by using this correlation, we can say which of option combination is real.
```

Comment:
I would also suggest to see the attribute constraint technique used by commercetools in order to check the validity of variant combination. That maybe a better logical approach to achieve this.

https://wiki.corp.adobe.com/display/EntComm/Product+Catalog+-+Competitors+Landscape#ProductCatalogCompetitorsLandscape-CommerceTools

https://docs.commercetools.com/merchant-center/product-types

Reply:
product types are entities more similar to Magento attribute sets, and as for me, this is something that we do have right now in M2 but do want to have in the storefront, because of a fact that options can not belong to the template in the most cases. As usual, if you have two products with the same list of options and options values that means that you have a duplicate of a product in your catalog.

Reply:
I think you are thinking in terms of Magento Options and Attributes, the way product type and attributes are used in commercetools is quite different, specially how attributes are helping to pave the way for the formation of variants; so the two concepts are merged together beautifully.

Reply:
that's exactly what https://docs.magento.com/user-guide/catalog/product-create-configurable.html does